pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,950 @@
+	// always be advertised.
+	if noLocalTrafficPolicy(svc) {
+		return
+	}
+	if c.hasAffectedBPByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice UPDATE event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) deleteEndpointSlice(obj interface{}) {
+	eps := obj.(*discovery.EndpointSlice)
+	svc, _ := c.serviceLister.Services(eps.GetNamespace()).Get(eps.GetLabels()[discovery.LabelServiceName])
+	if svc == nil {
+		return
+	}
+	// Events of EndpointSlices for Services without a `Local` traffic policy are ignored, as the Service IPs will
+	// always be advertised.
+	if noLocalTrafficPolicy(svc) {
+		return
+	}
+	if c.hasAffectedBPByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice DELETE event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) hasAffectedBPByEgress() bool {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if !c.matchedCurrentNode(bp) {
+			continue
+		}
+		if bp.Spec.Advertisements.Egress != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) addEgress(obj interface{}) {
+	eg := obj.(*v1beta1.Egress)
+	if eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if c.hasAffectedBPByEgress() {
+		klog.V(2).InfoS("Processing Egress ADD event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateEgress(oldObj, obj interface{}) {
+	oldEg := oldObj.(*v1beta1.Egress)
+	eg := obj.(*v1beta1.Egress)
+	if oldEg.Status.EgressNode != c.nodeName && eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if oldEg.Status.EgressIP == eg.Status.EgressIP && oldEg.Status.EgressNode == eg.Status.EgressNode {
+		return
+	}
+	if c.hasAffectedBPByEgress() {
+		klog.V(2).InfoS("Processing Egress UPDATE event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) deleteEgress(obj interface{}) {
+	eg := obj.(*v1beta1.Egress)
+	if eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if c.hasAffectedBPByEgress() {
+		klog.V(2).InfoS("Processing Egress DELETE event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) hasAffectedBPByNode(node *corev1.Node) bool {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if matchedNode(node, bp) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) addNode(obj interface{}) {
+	node := obj.(*corev1.Node)
+	if node.GetName() != c.nodeName {
+		return
+	}
+	if c.hasAffectedBPByNode(node) {
+		klog.V(2).InfoS("Processing Node UPDATE event", "Node", klog.KObj(node))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateNode(oldObj, obj interface{}) {
+	oldNode := oldObj.(*corev1.Node)
+	node := obj.(*corev1.Node)
+	if node.GetName() != c.nodeName {
+		return
+	}
+	if reflect.DeepEqual(node.GetLabels(), oldNode.GetLabels()) &&
```

Comment:
Why do we check labels and annotations only? it reminds me that the case for Flexible IPAM, Do we plan to support this? I suppose it's a valid case for BGP? 
@tnqn @antoninbas

Reply:
I may be wrong but my understanding is Flexible IPAM is another way to address the problem BGP tries to resolve, by dividing the Node network into multiple flat VLAN subnets and routing between them via a router, so BGP is not required here.

It needs to check labels and annotations because they affect whether the policy applies to the node and the node router ID.

Reply:
I didn't think about it too much but it's an interesting question. I agree with Quan's observation for the current version of Antrea Flexible IPAM.
However, if we interpret "Flexible IPAM" as the ability to assign arbitrary IPs to Pods, without looking at the current implementation (which requires bridging mode), then maybe this is something that BGP can actually help achieve (when supported by the underlay) without requiring that the Node's transport interface be moved to the OVS bridge? What do you think?

Note: that would be a L3 approach to Flexible IPAM, not a L2 approach using VLANs

Reply:
I re-opened this conversation not because it impacts this PR, but because I think it's a good thing to discuss

Reply:
Right, we should decouple IPAM and implementation. I agree BGPPolicy can help achieve Flexible IPAM without briding mode. Perhaps we can track it via an issue. If my understanding is correct, it may require the BGP processes running on Nodes exchange routes, not only advertise routes?

Reply:
> If my understanding is correct, it may require the BGP processes running on Nodes exchange routes, not only advertise routes?

I would say probably, yes. And that's what the Calico BGP implementation does I think.
I don't know if it is strictly required. If we keep the current approach (Node IPAM + Flexible IPAM, with the flexibility to choose between the 2 for different workloads), then I can see how just disabling SNAT could be enough, _assuming that for Pods using Flexible IPs the default route is used and traffic gets to a router to which Flexible IPs were advertised_. But for more generic support, and for the ability to send traffic to the destination Node directly (e.g. when 2 Nodes are on the same subnet) I would say being able to learn routes over BGP is necessary.